device/simulation/tt_simulation_device.cpp

```diff
@@ -0,0 +1,248 @@
+}
+
+void tt_SimulationDevice::deassert_risc_reset() {
+    log_info(tt::LogEmulationDriver, "HELLO WORLD FROM DEASSERT_RISC_RESET");
+    auto wr_buffer = create_flatbuffer(DEVICE_COMMAND_ALL_TENSIX_RESET_DEASSERT, std::vector<uint32_t>(1, 0), {0, 0, 0}, 0);
+    uint8_t *wr_buffer_ptr = wr_buffer.GetBufferPointer();
+    size_t wr_buffer_size = wr_buffer.GetSize();
+
+    print_flatbuffer(GetDeviceRequestResponse(wr_buffer_ptr)); // sanity print
+    host.send_to_device(wr_buffer_ptr, wr_buffer_size);
+}
+
+void tt_SimulationDevice::deassert_risc_reset_at_core(tt_cxy_pair core) {
+    log_info(tt::LogEmulationDriver, "HELLO WORLD FROM DEASSERT_RISC_RESET_AT_CORE");
+    // Ignore core for now, not implemented on sim side
+    auto wr_buffer = create_flatbuffer(DEVICE_COMMAND_ALL_TENSIX_RESET_DEASSERT, std::vector<uint32_t>(1, 0), {0, 0, 0}, 0);
+    uint8_t *wr_buffer_ptr = wr_buffer.GetBufferPointer();
+    size_t wr_buffer_size = wr_buffer.GetSize();
+
+    print_flatbuffer(GetDeviceRequestResponse(wr_buffer_ptr)); // sanity print
+    host.send_to_device(wr_buffer_ptr, wr_buffer_size);
+}
+
+void tt_SimulationDevice::assert_risc_reset_at_core(tt_cxy_pair core) {}
+
+void tt_SimulationDevice::close_device() {
+    // disconnect from remote connection
+}
+
+// Runtime Functions
+void tt_SimulationDevice::write_to_device(const void *mem_ptr, uint32_t size_in_bytes, tt_cxy_pair core, uint64_t addr, const std::string& tlb_to_use, 
+                    bool send_epoch_cmd, bool last_send_epoch_cmd, bool ordered_with_prev_remote_write) {
+    log_info(tt::LogEmulationDriver, "Device writing");
+    std::vector<std::uint32_t> data((uint32_t*)mem_ptr, (uint32_t*)mem_ptr + size_in_bytes / sizeof(uint32_t));
+    auto wr_buffer = create_flatbuffer(DEVICE_COMMAND_WRITE, data, core, addr);
+    uint8_t *wr_buffer_ptr = wr_buffer.GetBufferPointer();
+    size_t wr_buffer_size = wr_buffer.GetSize();
+    
+    print_flatbuffer(GetDeviceRequestResponse(wr_buffer_ptr)); // sanity print
+    host.send_to_device(wr_buffer_ptr, wr_buffer_size);
+}
+
+void tt_SimulationDevice::read_from_device(void* mem_ptr, tt_cxy_pair core, uint64_t addr, uint32_t size, const std::string& fallback_tlb) {
+    // std::vector<uint32_t> rd_resp;
+    // read_from_device(rd_resp, core, addr, size, fallback_tlb);
+    void *rd_resp;
+    size_t rd_rsp_sz;
+
+    // Send read request
+    auto rd_req_buf = create_flatbuffer(DEVICE_COMMAND_READ, {0}, core, addr, size);
+    host.send_to_device(rd_req_buf.GetBufferPointer(), rd_req_buf.GetSize());
+
+    // Get read response
+    host.recv_from_device(&rd_resp, &rd_rsp_sz);
+
+    auto rd_resp_buf = GetDeviceRequestResponse(rd_resp);
+    if (addr != 0x40){
+        log_info(tt::LogEmulationDriver, "Device reading vec");
+        print_flatbuffer(rd_resp_buf); // 0x40 is host polling device, don't print since it'll spam
+    }
+    std::vector<uint32_t> rd_resp_vec(rd_resp_buf->data()->begin(), rd_resp_buf->data()->end());
```

Comment:
Is there a way to do this without the allocation and copy into the vector?

Reply:
`std::memcpy(mem_ptr, rd_resp_buf->data()->data(), rd_resp_buf->data()->size() * sizeof(uint32_t));` works I believe. If that's what you mean